README.md
# yaqd-scpi

[![PyPI](https://img.shields.io/pypi/v/yaqd-scpi)](https://pypi.org/project/yaqd-scpi)
[![Conda](https://img.shields.io/conda/vn/conda-forge/yaqd-scpi)](https://anaconda.org/conda-forge/yaqd-scpi)
[![yaq](https://img.shields.io/badge/framework-yaq-orange)](https://yaq.fyi/)
[![black](https://img.shields.io/badge/code--style-black-black)](https://black.readthedocs.io/)
[![ver](https://img.shields.io/badge/calver-YYYY.0M.MICRO-blue)](https://calver.org/)
[![log](https://img.shields.io/badge/change-log-informational)](https://gitlab.com/yaq/yaqd-scpi/-/blob/main/CHANGELOG.md)

yaq daemons for SCPI hardware.
This package contains a few very generic daemons for interacting with the SCPI bus from yaq.
The generic approach works well for many simple applications, but more complex hardware interaction may require a more hardware-specific daemon.

## scpi-sensor

[`scpi-sensor`](https://yaq.fyi/daemons/scpi-sensor) allows yaq users to treat one or more scalar queries as a standard yaq sensor.
Multiple queries are treated as multiple channels.
Users can define channel names and units via config.
An example config follows:

```toml
[rigol]
port = 39999
visa_address = "TCPIP0::128.104.68.66::INSTR"
[rigol.channels]
[rigol.channels.trigger_level]
query = "TRIG:EDGE:LEV?"
units = "V"
[rigol.channels.time_offset]
query = "TIM:MAIN:OFFS?"
units = "s"
```

## scpi-set-continuous

[`scpi-set-continuous`](https://yaq.fyi/daemons/scpi-set-continuous) allows yaq users to address a single scalar settable.
Limits and units can be provided via config.
An example config follows:

```toml
[rigol_ch1_freq]
port = 39998
visa_address = "TCPIP0::128.104.68.66::INSTR"
limits = [0.1, 25e9]
scpi_command = "SOUR1:FREQ"

[rigol_ch2_freq]
port = 39999
visa_address = "TCPIP0::128.104.68.66::INSTR"
limits = [0.1, 25e9]
scpi_command = "SOUR1:FREQ"
```

## scpi-set-discrete

[`scpi-set-discrete`](https://yaq.fyi/daemons/scpi-set-discrete) allows yaq users to address a single non-scalar settable according to user-friendly names.
Identifiers must be provided via config.
An example config follows:

```toml
[rigol_ch1_func]
port = 39999
visa_address = "TCPIP0::128.104.68.66::INSTR"
scpi_command = "SOUR1:FUNC"
identifiers = {"SIN"=0, "SQU"=1, "RAMP"=2, "PULS"=3, "NOIS"=4, "DC"=5}
```

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/).

## [Unreleased]

### Changed
- rerendered avpr corresponding to yaq-traits 2022.5.0

## [2021.11.0]

### Fixed
- bad imports in sensor daemon
- discrete daemon bug where get_identifier was broken

## [2021.10.0]

### Changed
- rerender avpr based on recent traits change

## [2021.3.0]

### Added
- initial release

[Unreleased]: https://gitlab.com/yaq/yaqd-scpi/-/compare/v2021.11.0...main
[2021.11.0]: https://gitlab.com/yaq/yaqd-scpi/-/compare/v2021.10.0...v2021.11.0
[2021.10.0]: https://gitlab.com/yaq/yaqd-scpi/-/compare/v2021.3.0...v2021.10.0
[2021.3.0]: https://gitlab.com/yaq/yaqd-scpi/-/tags/v2021.3.0


yaqd_scpi/_scpi_sensor.py
__all__ = ["SCPISensor"]

import sys
import asyncio
from typing import Dict, Any, List

import pyvisa

from yaqd_core import HasMeasureTrigger, IsSensor, IsDaemon


class SCPISensor(HasMeasureTrigger, IsSensor, IsDaemon):
    _kind = "scpi-sensor"

    def __init__(self, name, config, config_filepath):
        super().__init__(name, config, config_filepath)
        self._channel_names = list(config["channels"].keys())
        self._channel_units = {k: config["channels"][k]["units"] for k in self._channel_names}
        if sys.platform.startswith("win32"):
            rm = pyvisa.ResourceManager()  # use ni-visa backend
        else:
            rm = pyvisa.ResourceManager("@py")  # use pyvisa-py backend
        self._instrument = rm.open_resource(config["visa_address"])

    async def _measure(self):
        out = {}
        for k in self._channel_names:
            query = self._config["channels"][k]["query"]
            self._instrument.write(query)
            out[k] = float(self._instrument.read())
        return out


yaqd_scpi/_scpi_set_continuous.py
__all__ = ["SCPISetContinuous"]

import sys
import asyncio
from typing import Dict, Any, List

import pyvisa

from yaqd_core import HasLimits, HasPosition, IsDaemon


class SCPISetContinuous(HasLimits, HasPosition, IsDaemon):
    _kind = "scpi-set-continuous"

    def __init__(self, name, config, config_filepath):
        super().__init__(name, config, config_filepath)
        self._scpi_command = self._config["scpi_command"]
        if sys.platform.startswith("win32"):
            rm = pyvisa.ResourceManager()  # use ni-visa backend
        else:
            rm = pyvisa.ResourceManager("@py")  # use pyvisa-py backend
        self._instrument = rm.open_resource(config["visa_address"])

    def _set_position(self, position):
        self._busy = True
        self._instrument.write(f"{self._scpi_command} {position}")

    async def update_state(self):
        while True:
            response = self._instrument.query(f"{self._scpi_command}?").strip()
            self._state["position"] = float(response)
            if self._state["position"] == self._state["destination"]:
                self._busy = False
                await asyncio.sleep(0.01)  # fast polling to release busy ASAP during orchestration
            else:
                await asyncio.sleep(0.1)  # slow polling for out-of-band input


yaqd_scpi/_scpi_set_discrete.py
__all__ = ["SCPISetDiscrete"]

import asyncio
import sys
from typing import Dict, Any, List

import pyvisa

from yaqd_core import IsDiscrete, HasPosition, IsDaemon


class SCPISetDiscrete(IsDiscrete, HasPosition, IsDaemon):
    _kind = "scpi-set-discrete"

    def __init__(self, name, config, config_filepath):
        super().__init__(name, config, config_filepath)
        self._scpi_command = self._config["scpi_command"]
        self._identifiers_by_position = {k: v for v, k in self._config["identifiers"].items()}
        if sys.platform.startswith("win32"):
            rm = pyvisa.ResourceManager()  # use ni-visa backend
        else:
            rm = pyvisa.ResourceManager("@py")  # use pyvisa-py backend
        self._instrument = rm.open_resource(config["visa_address"])

    def _set_position(self, position):
        self._busy = True
        identifier = self._identifiers_by_position[position]
        self._instrument.write(f"{self._scpi_command} {identifier}")

    async def update_state(self):
        while True:
            response = self._instrument.query(f"{self._scpi_command}?").strip()
            self._state["position_identifier"] = response
            position = self._config["identifiers"][response]
            self._state["position"] = position
            if self._state["position"] == self._state["destination"]:
                self._busy = False
                await asyncio.sleep(0.01)  # fast polling to release busy ASAP during orchestration
            else:
                await asyncio.sleep(0.1)  # slow polling for out-of-band input
